add module aws_step_functions_state_machine_execution

prasadkatti opened the pull request:
##### SUMMARY
New module to start or stop execution of an AWS Step functions state machine

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
aws_step_functions_state_machine_execution

##### ADDITIONAL INFORMATION
- The module allows starting or stopping an execution of a state machine. The scope does not include listing or describing executions.
- I will add a PR to the aws-terminator project as well. (https://github.com/mattclay/aws-terminator/pull/71)

prasadkatti committed:
add module aws_step_functions_state_machine_execution

prasadkatti commented:
ready_for_review

tremble commented on lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:
```diff
@@ -0,0 +1,178 @@
+
+options:
+    action:
+        description: Desired action (start or stop) for a state machine execution
```
niggle:

```suggestion
        description: Desired action (start or stop) for a state machine execution.
```

Descriptions should end with a "."
https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-block

tremble commented on test/integration/targets/aws_step_functions_state_machine/tasks/main.yml:
```diff
@@ -130,6 +130,74 @@
+        name: "{{ execution_name }}"
+        execution_input: "{}"
+        state_machine_arn: "{{ creation_output.state_machine_arn }}"
+        <<: *aws_connection_info
```
Niggle 2:

The common AWS connection info can be better written by adding module_defaults to the top level block:

```
- name: Integration test for AWS Step Function state machine module
  module_defaults:
    group/aws:
      aws_access_key: "{{ aws_access_key }}"
      aws_secret_key: "{{ aws_secret_key }}"
      security_token: "{{ security_token | default(omit) }}"
      region: "{{ aws_region }}"
  block:
```

This means you can then remove all of the ``<<: *aws_connection_info`` lines and the "Set connection information for all tasks" task.

tremble replied:
Oh, and related niggle 3, please add the new module to lib/ansible/config/module_defaults.yml

prasadkatti replied:
done. Thanks for the review.

tremble reviewed (changes requested):
Couple of minor niggles...

jillr commented on lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:
```diff
@@ -0,0 +1,178 @@
+        module.fail_json_aws(e, msg="Failed to start execution.")
+
+    module.exit_json(changed=True, execution_arn=execution.get('executionArn'),
+                     start_date=execution.get('startDate'))
```
You can achieve this same result output with
`module.exit_json(changed=True,**camel_dict_to_snake_dict(execution))`
which can be imported from `ansible.module_utils.ec2`. That will automatically cover if AWS ever starts returning additional keys from this API and is generally easier to maintain.

prasadkatti replied:
Thanks for the review. I have pushed a change for this.

jillr commented on lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:
```diff
@@ -0,0 +1,178 @@
+    except ClientError as e:
+        module.fail_json_aws(e, msg="Failed to stop execution.")
+
+    module.exit_json(changed=True, stop_date=res.get('stopDate'))
```
```suggestion
    module.exit_json(changed=True,**camel_dict_to_snake_dict(res)))
```

jillr commented on lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:
```diff
@@ -0,0 +1,178 @@
+
+    action = module.params.get('action')
+    if action == "start":
+        start_execution(module, sfn_client)
```
`client.start_execution` is idempotent but `start_execution` always returns `changed=True` if `action=start`. We probably want to `client.list_executions` to see if the execution is already running before proceeding.

prasadkatti replied:
> The `list_executions` is eventually consistent. The results are best effort and may not reflect very recent updates and changes.

So we could potentially get false positives (changed=True) if the execution was already running but did not get listed by the `list_executions` call.

prasadkatti replied:
> StartExecution is idempotent. If StartExecution is called with the same name and input as a running execution, the call will succeed and return the same response as the original request. If the execution is closed or if the input is different, it will return a 400 ExecutionAlreadyExists error. Names can be reused after 90 days.

I return `changed=False` if ExecutionAlreadyExists is returned. That would mean that a new execution was not started. 

But I am not sure if we want to return `changed=False` in the case of a running execution. I guess that really depends on how the value of `changed` is going to be used further down the playbook. Also, if the AWS service isn't making the distinction clear maybe we don't need to either. What do you think?

jillr replied:
I haven't actually used stepfunctions myself so I'm just going off the docs. AIUI, `ExecutionAlreadyExists` would only be returned on a running execution if the `execution_input` is different, not in a case where the input is the same.  As an ansible user who is new to step functions, I feel like if I gave the same input and ran the playbook twice I would expect the 2nd time to be `changed=false` if the first execution is still running.  But that doesn't mean that's how AWS users familiar with step functions will expect it.

prasadkatti replied:
We could just document that the value of `changed` for the tasks will be aligned to how boto3 call returns. If we get `ExecutionAlreadyExists`, we return `False`. Otherwise we return `True`. Again, what the right value for `changed` is depends on how the user of the module wishes to use it. We should just document it precisely and then gather feedback from the users as they start using the module. Almost wish there was a thing like beta modules to gather early feedback.

prasadkatti replied:
Between ansible and AWS, we should be more aligned with AWS on this one, IMO. If they don't make a distinction between `ExecutionAlreadyExists` and `ExecutionCurrentlyRunning`, then we shouldn't either. This and the fact that `list_executions` is only eventually consistent makes me think the current behavior is what we want.

jillr replied:
That makes sense, and I definitely agree we should very clearly document the behaviour.
@tremble, any thoughts on this approach to idempotency here before we proceed?

tremble replied:
I have a *slight* preference for @jillr's approach, and trying to ensure that check_mode behaves as you might expect.  But it's not a strong enough preference to hold anything up.  The behaviour should certainly be documented though.

prasadkatti replied:
So...? What should we do here? Document the behavior clearly? If that is what we plan to do I can push another patch.
Given that with the eventually consistent `list_executions` you could get inconsistent results, do we still want to go with Ansible's definition for idempotency here?

s-hertel replied:
I like @jillr's approach. We should err on the side of trying to keep our AWS modules similar. There are a lot of AWS services that use eventual consistency so we have to work around it as best we can. Using waiters after making API calls can help a little bit, though I'm not sure that it's applicable here (I see that there are no boto3 built-in waiters yet for this service, but you could possibly add your own criteria for waiting to https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/aws/waiters.py). In some circumstances we've added brief pause tasks to integration tests to help. Do add a note in the documentation about the eventual consistency problem.

prasadkatti replied:
I have pushed more changes. For start execution, we now list all executions and check if there is an execution with the same name. If there is one, the module exits with `changed=False`. For stop execution, we describe the execution and exit with `changed=False` unless the execution status is '_RUNNING_'. I have also added tests around these changes and comments wherever I felt necessary.

jillr commented on test/integration/targets/aws_step_functions_state_machine/tasks/main.yml:
```diff
@@ -130,6 +130,74 @@
+          - start_execution_output.changed
+          - "'execution_arn' in start_execution_output"
+          - "'start_date' in start_execution_output"
+
```
Once the module is idempotent, we'll want another test like `name: Start execution of state machine - idempotent`, then assert that output is not changed.

prasadkatti replied:
added more tests for idempotency and check mode

jillr commented on test/integration/targets/aws_step_functions_state_machine/tasks/main.yml:
```diff
@@ -130,6 +130,74 @@
+        that:
+          - stop_execution_output.changed
+          - "'stop_date' in stop_execution_output"
+
```
And an idempotent stop test.

prasadkatti replied:
done 👍

jillr reviewed (commented):
Thanks @prasadkatti, this module looks pretty good just a couple things.

prasadkatti committed:
AWS step functions tests - Use module defaults

prasadkatti committed:
Return all attributes from aws api calls as ansible task output

s-hertel commented on lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:
```diff
@@ -0,0 +1,178 @@
+        if e.response['Error']['Code'] == 'ExecutionAlreadyExists':
+            module.exit_json(changed=False)
+        module.fail_json_aws(e, msg="Failed to start execution.")
+
```
Please fail safely for the module if a BotoCoreError exception happens here too and in stop_execution.

prasadkatti replied:
fixed, thanks for the review

prasadkatti committed:
aws_sfn - make start and stop execution idempotent and fix check mode

prasadkatti commented:
ready_for_review
Tests failing on AccessDenied to some actions added in https://github.com/mattclay/aws-terminator/pull/80

jillr commented on lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:
```diff
@@ -103,40 +103,60 @@
+        # list_executions is eventually consistent
+        resp_iterators = sfn_client.get_paginator('list_executions').paginate(stateMachineArn=state_machine_arn)
+
+        for resp in resp_iterators:
```
Paginators have a `build_full_result` method you can use, which is a little nicer than nested loops. So you could do:
```
current_executions = resp_iterators.build_full_result()['executions']
for execution in current_executions:
    if name == execution['name']:
```

prasadkatti replied:
Thanks, I will make the required changes. Does it look good otherwise?

prasadkatti committed:
aws sfn - use build_full_result method of the paginator

jillr commented on lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py:
```diff
@@ -0,0 +1,197 @@
+        for execution in page_iterators.build_full_result()['executions']:
+            if name == execution['name']:
+                check_mode(module, msg='State machine execution already exists.', changed=False)
+                module.exit_json(changed=False, something="something")
```
Sorry I missed this before @prasadkatti, did you mean to put a msg here other than 'something'?

prasadkatti replied:
Oops. Sorry about that. I was just using that extra return parameter as a way of distinguishing between the 2 `exit_json`s and making sure the right one was called. I will remove it right now.

BTW, the `ExecutionAlreadyExists` block might be unreachable now. But I left in because if that exception does somehow get returned, we don't want the task to fail.

jillr reviewed (commented):
One last small thing that I missed before, then this looks good.

prasadkatti committed:
aws sfn - remove changes made to help with local debugging

jillr reviewed (approved):
LGTM once terminator change is merged

jillr commented:
rebuild_merge

prasadkatti commented:
Hi @jillr, ansibot did not pick up the `rebuild_merge`. Could it be because the PR has `needs_ci_update` and `needs_revision` labels?

This is the log from (https://ansibullbot.eng.ansible.com/cgi-bin/ansibot_status.cgi)

> 2019-12-06 00:34:07,690 INFO skipping: no changes since last run
2019-12-06 00:34:07,691 INFO finished triage for https://github.com/ansible/ansible/pull/64431
2019-12-06 00:34:07,691 INFO starting triage for https://github.com/ansible/ansible/pull/64428
2019-12-06 00:34:08,095 INFO skipping: no changes since last run

tremble commented:
The needs_revision is because the last test run failed.  Simplest option is probably to rebase and force push to see if the build works now.

prasadkatti commented:
Closed and reopened the PR to kick off the CI run.

prasadkatti commented:
The failed tests are not related to the PR - https://app.shippable.com/github/ansible/ansible/runs/152945/112/tests

jillr commented:
F31 tests have been marked unstable, it's unrelated to this PR. 65671
